ginga/canvas/types/plots.py

```diff
@@ -0,0 +1,832 @@
+
+    def update_resize(self, viewer, dims):
+        """Called when the viewer is resized."""
+        self.recalc(viewer)
+
+    def update_elements(self, viewer):
+        """This method is called if the plot is set with new points,
+        or is scaled or panned with existing points.
+        """
+        #self.recalc(viewer)
+        pass
+
+    def get_latest(self):
+        """Get the latest (last) point on the plot.  Returns None if there
+        are no points.
+        """
+        if len(self.points) == 0:
+            return None
+        return self.points[-1]
+
+    def get_limits(self, lim_type):
+        """Get the limits of the data or the visible part of the plot.
+
+        If `lim_type` == 'data' returns the limits of all the data points.
+        Otherwise returns the limits of the visibly plotted area.  Limits
+        are returned in the form ((xmin, ymin), (xmax, ymax)), as an array.
+        """
+        if lim_type == 'data':
+            # data limits
+            return np.asarray(self.limits)
+
+        # plot limits
+        self.path.crdmap = self.crdmap
+        if len(self.path.points) > 0:
+            llur = self.path.get_llur()
+            llur = [llur[0:2], llur[2:4]]
+        else:
+            llur = [(0.0, 0.0), (0.0, 0.0)]
+        return np.asarray(llur)
+
+    def draw(self, viewer):
+        """Draw the plot.  Normally not called by the user, but by the viewer
+        as needed.
+        """
+        self.path.crdmap = self.crdmap
+
+        self.recalc(viewer)
+
+        if len(self.path.points) > 0:
+            self.path.draw(viewer)
+
+
+class Axis(CompoundObject):
+    """
+    Base class for axis plotables.
+    """
+    def __init__(self, title=None, num_labels=4, font='sans',
+                 fontsize=10.0):
+        super(Axis, self).__init__()
+
+        self.aide = None
+        self.num_labels = num_labels
+        self.title = title
+        self.font = font
+        self.fontsize = fontsize
+        self.grid_alpha = 1.0
+        self.format_value = self._format_value
+
+    def register_decor(self, aide):
+        self.aide = aide
+
+    def _format_value(self, v):
+        """Default formatter for XAxis labels.
+        """
+        return "%.4g" % v
+
+    def set_grid_alpha(self, alpha):
+        """Set the transparency (alpha) of the XAxis grid lines.
+        `alpha` should be between 0.0 and 1.0
+        """
+        for i in range(self.num_labels):
+            grid = self.grid[i]
+            grid.alpha = alpha
+
+    def get_data_xy(self, viewer, pt):
+        arr_pts = np.asarray(pt)
+        x, y = viewer.tform['data_to_plot'].from_(arr_pts).T[:2]
+        flips = viewer.get_transforms()
+        if flips[2]:
+            x, y = y, x
+        return (x, y)
+
+    def get_title(self):
+        titles_d = self.aide.get_axes_titles()
+        return titles_d[self.kind]
+
+    def add_plot(self, viewer, plot_src):
+        pass
```

Comment:
Why `pass`? Why not `raise NotImplementedError`?

Reply:
Good point.

Reply:
Hmm, I guess one reason is that ``XAxis`` and ``YAxis`` then have to reimplement the methods with ``pass`` if they don't need to update their state when a plot is added or deleted.  Neither of them need to do anything for this part of the internal plot decor API.  I don't know if it makes it more clear to duplicate these methods in the subclasses.  I changed it so that there is a more detailed comment above the ``pass``.